Domain Model (#1554)
thin-edge/thin-edge.io: docs/src/architecture/domain-model.md

The **cloud** could be some IoT Cloud Platform, such as [Cumulocity](https://www.softwareag.cloud/site/product/cumulocity-iot.html), [Azure](https://azure.microsoft.com) or [AWS](https://aws.amazon.com); or an on-premise "IoT system".

**thin-edge** has been designed with the assumption that the **device** is represented in the **IoT cloud** as **device twin**.
The **device twin** in the IoT cloud is a logical representation of the **device**. As such, it carries or references **Telemetry Data**, and stores the _state_ of the **device**.
Here _state_ refers to e.g. **Device Management** details as _installed software versions_ or _configurations_.

## Telemetry Data

**Telemetry data** are _measurements_, _events_ and _alarms_.

  * **Measurement**:
     * has a _type name_
     * carries a _series_ of **samples**
       * a **sample** is a single value or set of values, all taken at a single point in time
         * could be a mix of values coming from physical **Sensors**[^1] and the device's **Domain Application**[^1];
           e.g. _voltage_ and _current_ of an electricity meter, and current state of the manufacturing control process

>          * all **sample's** values could be a mix of numbers, strings or booleans

Author: Indeed, that is not a typo. We want to get rid of that limitation to numbers.

         * a **sample** has _one_ timestamp
       * a _series_ of **samples** is a set of **samples** over a period of time;
         e.g. _voltage_ and _current_ captured in a fixed interval
     * a measurement optionally holds individual _units_ for the values of the **samples**
     * a measurement relates to a source, i.e. a device or a process

>   * **Setpoint**, is a single value or set of values

Author: The values are process variables, used by the Domain Application.

I added that information, see https://github.com/cstoidner/thin-edge.io/commit/6e55451f90c2594869e3b5aaf7430a080fcbe1cb

    * values are visible in the cloud, similar to measurements (e.g. a "temperature upper limit")
    * values are used on the device-site, e.g. the **Domain Application**[^1]
    * values can be manipulated from cloud-site or the device-site, e.g. to
      - set one or more process variables[^1] (e.g. upper/lower limits or threshold of a climate control)

>     * a manipulation leads not to an synchronous action (like an operation),

Author: I feel the distinction is now clearer with that change: https://github.com/cstoidner/thin-edge.io/commit/6e55451f90c2594869e3b5aaf7430a080fcbe1cb

If you see still need for improvement, please tell so.

      instead new values are used by the **Domain Application**[^1] asynchronously from next control-cycle on
    * values could be a mix of numbers, strings or booleans
  * **Event**, is a notification that something happened on the device's environment or software system

>     * its source could be

Author: Thats true, using the term _source_ here does not convey what I was trying to say. I improved now that part, see commits https://github.com/thin-edge/thin-edge.io/commit/704bf5aa0ea5113dbf0a7b535ef992d20b67f624 and https://github.com/thin-edge/thin-edge.io/commit/26c20ee7ea8b524fe4458206a08f1b002680e745

      * the **domain application**[^1]
      * parts from **OS / Libs / Runtime**[^1], or the whole thing as one
      * **thin-edge** itself
* **Configuration Management**:
    * allows viewing and changing configurations on the device, as
      * list available configurations
      * transfer individual configurations from device to cloud, and vice versa
    * a configuration is a text file or a binary file, as e.g.
      * configuration file(s) of the **domain application**[^1]
      * configuration file(s) of **OS / Libs / Runtime**[^1]
      * configuration file(s) of **thin-edge** it-self
* **Log Management**:
    * allows viewing logs from the device
      * lists available logs
      * transfer individual logs from device to cloud
    * a log is a text file, as e.g.
      * log file(s) of the **domain application**[^1]
      * log file(s) of **OS / Libs / Runtime**[^1]
      * log file(s) of **thin-edge** it-self
* **Device Monitoring**:
    * collects metrics (e.g. cpu, memory and disk metrics) from the device and forward these to the IoT cloud
    * allows monitoring the health of devices
    * helps troubleshooting when problems with the device are reported

> 

Author: Firmware management is now added, see https://github.com/thin-edge/thin-edge.io/pull/1554/commits/1f1678ae726a4279a98fa72eb8949026b2bcc8a3

NOTE: @reubenmiller  and me decided to consider "device profiles" in a second step. And in that current domain model approach there is nothing that stops us to put "device profiles" later on top of that.

  * the device **thin-edge** is running on is referred as the **main-device**
    * **thin-edge** on the **main-device** establishes and manages all communication to the cloud
  * all devices connected to the **main-device** are referred as **external child-devices** 
  * each **external child-device** can be represented in the cloud with its individual **device twin**
    * a unique **child-id** makes the association between each **external child-device** and it's **device twin**
  * all **telemetry data** and **device management** functionality can appear in the context of the 
    **external child-device's** individual **device twin**, or the **main-device's** **device twin**
  * also _containers_ or _processes_ running on the **main-devices** can be handled like **external child-device**;
    those are referered then as **logical child-devices**

The figure below illustrates the child-device concept.

![Device Concept](images/device-concept.svg)

## Plugins

**thin-edge** realizes **Device Management** based on **plugins**.
  * a **plugin** can be an (external) executable (e.g. as the `c8y_configuration_plugin` for _configuration management_)

>     or a thin-edge built-in software component (e.g. as for _software management_)

Author: > or a thin-edge built-in software component (e.g. `tedge-agent` for _software management_)

`tedge-agent` does more than just _software management_. So I slightly changed your suggestion. Whats about that: (?) 

```suggestion
    or part of a thin-edge built-in software component (e.g. _software management_ in `tedge-agent`)
```

  * to realize **Device Management** functionality, a **plugin** accesses _resources_ and _services_ of the device, as e.g.
      * _software management_ uses the device's _package manager_
      * _configuration management_ reads/writes device's _configuration files_
  * a **plugin** can facilitate **Device Management** functionality for the **main-device** and **external child-device**, but runs usually only on the **main-device**

>     * _resources_ and _services_ (e.g. file system, package manager, ...) of the **main-device** can be directly accessed by the **plugin**, as it runs on the **main-device**

Author: That was not meant as restriction, however I see that wording is missleading.

So I highlighted now, that _child-device agents_ are also the way for containerized environments. See https://github.com/thin-edge/thin-edge.io/commit/f3d14c12738adcdd924aaff76a462596be94d9d3

    * all _resources_ and _services_ of an **external child-device** are only remotely accessible for the **main-device**

>       * the **plugin** running on the **main-device** needs another software component that establish the remote access for those _resources_ and _services_

Author: The "other software component" (i.E. the child-device agent) must not necesarilly run on the child-device.

As you can see here (https://github.com/thin-edge/thin-edge.io/pull/1554/files/7ef7c3efe7ed174220c16c1415786eb3052ac7cb#diff-8fb5e0172ee785582c478280a6ef2f51c91d90961ee5f799ba1f5b212ae57c99R130) it can installed and executed on the **external child-device** or the **main-device**.

      * that software component is referred as **child-device agent**
  * a **plugin** can also facilitate **Device Management** functionality for containers running on the **main-device**

>     * similar to **external child-devices**, all _resources_ and _services_ of a container are accessed by a **plugin** via a **child-device agent** that runs inside the corresponding container 

Author: Having in mind that a container is a **logical child-device**, it might not sound so weird. But I agree, for one who has strong focus on containers and does not deal with child-devices at all, it might be irritating.

However I was already from beginning not really fine with the term **... "agent"**. And now with **plugin agent** it gives even less a clue what that thing is about.

What do you think about the term **plugin adapter**? In software an adapter "provides a compatible connection between the components of a system". And that is exactly what our piece of software is doing.

@albinsuresh , what do you think about that?

>     * similar to **external child-devices**, all _resources_ and _services_ of a container are accessed by a **plugin** via a **child-device agent** that runs inside the corresponding container 

Author: I renamed now "child device agent" to "plugin adapter".

See https://github.com/thin-edge/thin-edge.io/pull/1554/commits/c75a26ddc532bd3ef547bcc1071feaf7ef047e28

### Child-Device Agent
  * a **child-device agent** is the counterpart of a **plugin**, that takes the responsibility to access to the **external child-device's** _resources_
  * a **child-device agent** can serve one or more **plugins**
  * a **child-device agent** can be installed and executed on the **external child-device**, or on the **main-device**
    * if it runs in the **external child-device** it can access the _resources_ directly
    * if it runs on the **main-device** it can use any (low-level) interfaces the **external child-device** provides to access those resources
      * One main reason to install the **child-device agent** on the **main-device** is, when the **external child-device** cannot or shall not be altered.

### Plugin-Contract

A **plugin** defines and implements a specific **contract** for all interactions with a **child-device agent**
  * part of the **contract** could be e.g.:
      * the **child-device agent** must listen and react to certain requests of the **plugin**, e.g. on MQTT
      * the **child-device agent** must provide/consume files to/from the **plugin** on purpose, e.g. via HTTP
      * ...and more...

>   * a **plugin's** **contract** can be flagged with a unique name (e.g. `tedge_config`)

Reviewer: How about calling this unique name a `plugin-capability` instead of `plugin-contract`. A `plugin-capability`.

Author: True, "plugin-contract name" doesn't sound perfect. 

> How about calling this unique name a plugin-capability instead of plugin-contract.

I think term "plugin-capability" is also not the right thing. Cause the "contract" is not a capabiity, but instead the whole plugin might be more a "capability".

However to address you point I introduced now the term "plugin-identifier" as name for the whole plugin, and removed the "plugin-contract name". See https://github.com/thin-edge/thin-edge.io/pull/1554/commits/98bd39eb0c536930363ee7e271da21598e0b1b98

I feel that improves that section a lot.

      * ...and more...
  * a **plugin** can be flagged with a unique _plugin-identifier_ (e.g. `tedge_config`, `tedge_software` or any other unique string for a custom specific plugin)
    * based on that _plugin-identifier_ a **plugin adapter** can report and find **plugins** the child-device intends to contact (e.g. during provisioning phase)
    * a **plugin** and a **plugin adapter** using the same _plugin-identifier_ can assume to implement the same **contract**, and to be compatible with each other

## Inventory

**thin-edge** holds and manages an **inventory** on the **main-device**, that stores and provides information about the **main-device** and known **child-devices**
  * information stored per **device** are
    * supported kinds of **Device Management** capabilities/operations
    * supported kinds of **Telemetry Data**
    * optionally any custom-specific meta-information per **device**
  * the **inventory** is the communication backbone for **plugins**, **external child-devices**, the **domain application**[^1] and **thin-edge** it-self
      * one can add information to announce capabilities a **device** supports

>       * another one can retrieve those information to identify capabililties a **device** supports

Author: Thats true. Same for the new telemetry item `setpoint` and telemetry details `units` and `samples`; and finally new device management functionality `firmware management`.

For all those development work need to be planned.

  * **Measurement**:
     * has a _type name_

>      * carries a _sequence_ of **samples**

Reviewer 2: The correct word is [time series](https://en.wikipedia.org/wiki/Time_series). A sequence is indexed by integers. A time-series by timestamps.

Author: Yes. The point is, I got the suggestion to avoid term "series" here, to prevent confusions with C8Y meaning of series. However I will reconsider that.

And finally, _time series of samples_ (by timestamps) is exactly what I meant here.

Author: Change to "time-series", see https://github.com/thin-edge/thin-edge.io/pull/1554/commits/bb58d5a1784b3b8e144704de8c2f41b5d71c9280
